src/core/api/media_element_track_choice_manager.ts

```diff
@@ -0,0 +1,468 @@
+      const language = normalizedLanguages[i];
+      for (let j = 0; j < this._audioTracks.length; j++) {
+        const audioTrack = this._audioTracks[j];
+        const normalizedLanguage = audioTrack.track.normalized;
+        if (normalizedLanguage === language) {
+          this.setAudioTrackById(audioTrack.track.id);
+          return;
+        }
+      }
+    }
+  }
+
+  private _setInitialTextTrack() : void {
+    const preferredTextTracks = this._preferredTextTracks.getValue();
+    const normalizedLanguages = preferredTextTracks
+      .filter(
+        (textTrack): textTrack is { language : string;
+                                    closedCaption : boolean; } => textTrack !== null)
+      .map(({ language }) =>  normalizeLanguage(language));
+
+    for (let i = 0; i < normalizedLanguages.length; i++) {
+      const language = normalizedLanguages[i];
+      for (let j = 0; j < this._textTracks.length; j++) {
+        const textTrack = this._textTracks[j];
+        const normalizedLanguage = textTrack.track.normalized;
+        if (normalizedLanguage === language) {
+          this.setTextTrackById(textTrack.track.id);
+          return;
+        }
+      }
+    }
+  }
+
+  /**
+   * Monitor HTML5 tracks changes
+   */
+  private _onTrackChange(mediaElement: HTMLMediaElement): void {
+    const audioCallback = () => {
+      if (this._audioTracks !== undefined) {
+        for (let i = 0; i < this._audioTracks.length; i++) {
+          const { track, origTrack } = this._audioTracks[i];
+          if (origTrack.enabled) {
+            this.trigger("trackChange", { type: "audio",
+                                          track });
+            return;
+          }
+        }
+      }
+      this.trigger("trackChange", { type: "audio",
+                                    track: null });
+      return;
+    };
+    const textCallback = () => {
+      if (this._textTracks !== undefined) {
+        for (let i = 0; i < this._textTracks.length; i++) {
+          const { track, origTrack } = this._textTracks[i];
+          if (origTrack.mode === "showing") {
+            this.trigger("trackChange", { type: "text",
+                                          track });
+            return;
+          }
+        }
+      }
+      this.trigger("trackChange", { type: "text",
+                                    track: null });
+      return;
+    };
+    const videoCallback = () => {
+      if (this._videoTracks !== undefined) {
+        for (let i = 0; i < this._videoTracks.length; i++) {
+          const { track, origTrack } = this._videoTracks[i];
+          if (origTrack.selected) {
+            this.trigger("trackChange", { type: "video",
+                                          track });
+            return;
+          }
+        }
+      }
+      this.trigger("trackChange", { type: "video",
+                                    track: null });
+      return;
+    };
+    if (mediaElement.audioTracks !== undefined) {
```

Comment:
Why writing all this here when the other event handlers are all up in the constructor with the same conditions

Reply:
Done